Декоративные элементы всех страниц

mis3rably opened the pull request:
---
:mortar_board: [Декоративные элементы всех страниц](https://up.htmlacademy.ru/htmlcss/35/user/1976751/tasks/13)

:boom: https://htmlacademy-htmlcss.github.io/1976751-device-35/11/

mis3rably committed:
Стилизация декоративных элементов

mis3rably commented on styles/styles.css:
```diff
@@ -893,35 +1052,353 @@ body {
+.advantages-link:focus-visible {
+	outline: 2px solid #AF4FFF;
+}
+.main-header-page-navigation-link::after {
```
Для стилизации ссылок когда появляется подчёркивание снизу я хотел использовать border-bottom, но из-за letter-spacing у текста сбоку появляется пустое место и выходит не очень. Я сделал с помощью after псевдо-элемента, и задал ему ширину 96% от родителя. Получилось лучше, но всё же я думаю это неидеальный вариант. Как сделать по-другому?

pavel-niukalo replied:
Через псевдоэлемент хорошее решение, единственное, поскольку псевдоэлемент абсолютно спозиционирован и задан как псевдоэлемент ссылки, лучше дать именно ссылке `position: relative`. Поскольку абсолютно спозиционированный элемент ищет ближайшего родителя, у которого `position` отличный от `static`. И тогда можно дать 100% ширины, `left: 0`, в целом будет хорошо выглядеть

mis3rably replied:
Исправил

mis3rably commented on styles/styles.css:
```diff
@@ -893,35 +1052,353 @@ body {
+	z-index: 0;
+	counter-increment: slider-number;
+}
+.product-showcase-item::after {
```
Цифру с номером сладера я реализовал с помощью псевдо-элемента и `absolute` позиционирования от родителя. По какой-то причине ссылка Контакты которая находится над ним перестаёт реагировать на `hover` частично, если отодвинуть чуть ниже то работает нормально, но тогда идёт несоответсвие по макету. Почему так происходит и как это решить?

pavel-niukalo replied:
Какое-то странное поведение, видимо где-то происходит наложение, надо покопать и поиграть со стилями, позже смогу посмотреть. Пока как вариант, можно всему `header` дать `z-index: 1`, и это решит проблему

mis3rably replied:
Спасибо, помогло)

mis3rably commented on index.html:
```diff
@@ -14,68 +14,74 @@
+              </a>
+            </li>
+          </ul>
+            <a href="catalog.html" class="main-header-page-navigation-button-link state-default flex">Каталог товаров
```
Проверяя `focus-visible` состояние элемента, я заметил что рамка снизу пропала. Тоже самое происходит в пагинации в каталоге. Если добавлять паддинги то выходит не очень, да и наверное есть вариант правильнее?

pavel-niukalo replied:
Рамка снизу пропала, поскольку она уходит под `main`. Как варианты решения: 1) задать `z-index` кнопке, чтобы она явно была "выше" по слою чем `main` 2) задать шапке нижний паддинг, чтобы зарезервировать место для `outline`

mis3rably replied:
Помогло, спасибо, пагинации добавил каждой кнопке `z-index` при `focus-visible`

mis3rably commented on index.html:
```diff
@@ -193,9 +201,9 @@ <h2 class="visually-hidden">Перечень товаров.</h2>
-              <button class="nuances-controls" type="button">Доставка</button>
-              <button class="nuances-controls" type="button">Гарантия</button>
-              <button class="nuances-controls" type="button">Кредит</button>
+              <button class="nuances-controls main-link nuances-controls-current" type="button">Доставка</button>
```
Черную обводку вокруг текущей ссылки в слайдере я сделал с помощью псевдо-элемента. При проверке работы ` focus-visible` выделяется не вся обводка а только кнопка. Я честно уже не могу сообразить как реализовать этот элемент по-другому, помогите пожалуйста

pavel-niukalo replied:
Тут я бы предложил вот эту черную обводку делать не псевдоэлементом, а растягивать кнопку саму. То есть текущая кнопка у нас будет больше обычной и с черной заливкой и тд. В таком случае мы сможем спокойно ее стилизовать при фокусе

mis3rably replied:
Исправил, спасибо) Вроде бы так просто а в голову совсем не приходило

mis3rably commented on catalog.html:
```diff
@@ -136,131 +144,154 @@ <h2 class="visually-hidden">Список моноподов для селфи с
-                      width="360"
-                      height="380"
-                    >
+                  <a href="product.html" class="product-card-link">
```
Я долго пытался добиться чтобы фокус работал в соответствии с Стайлгайдом, сначала выделялись только картинка и текст а не весь элемент целиком, после я обернул всю карточку в ссылку и фокус пропал совсем)

mis3rably replied:
Проблема осталась, так и не смог найти решения

pavel-niukalo replied:
Кстати, если обнулить `outline` у ссылки при фокусу, то в целом выглядит по стайлгайду

![image](https://user-images.githubusercontent.com/52506235/199055617-40312453-eac3-4db9-94f1-191fad858603.png)


Вполне возможно, что проблема с фокусом была из-за того, что спека нам не говорит напрямую, можно ли вкладывать `div` в `a` и браузер мог по-своему добавить обработку этого и как-то сломать фокус

![image](https://user-images.githubusercontent.com/52506235/199055870-9ca72615-ed0b-42ad-9ed0-beaea4d86562.png)

mis3rably replied:
Исправил, теперь буду знать о таком

pavel-niukalo commented on catalog.html:
```diff
@@ -14,57 +14,65 @@
+            </li>
+          </ul>
+            <a href="catalog.html" class="main-header-page-navigation-button-link state-default flex">Каталог товаров
+              <button class="main-header-page-navigation-button state-default" type="button">
```
Кнопку нельзя вкладывать в ссылки

![image](https://user-images.githubusercontent.com/52506235/198581579-dc35af20-c6c3-4423-bfee-744c2d35b73d.png)

pavel-niukalo commented on index.html:
```diff
@@ -14,68 +14,74 @@
+            </li>
+          </ul>
+            <a href="catalog.html" class="main-header-page-navigation-button-link state-default flex">Каталог товаров
+              <button class="main-header-page-navigation-button state-default" type="button">
```
Здесь аналогично про запрет вложения кнопки в ссылки

pavel-niukalo commented on styles/styles.css:
```diff
@@ -229,9 +239,11 @@ body {
 	align-items: center;
-	justify-content: space-between;
+	justify-content: flex-end;
+	background-image: url(../images/green-tick.svg);
```
Для надежности лучше добавить кавычки во все `url` - `url("../")`

pavel-niukalo commented on styles/styles.css:
```diff
@@ -503,18 +528,62 @@ body {
+
+.login {
+	background-image: url(../images/login.svg);
+    background-repeat: no-repeat;
+    background-position-y: 9px;
+    background-size: 13px 12px;
+    padding-left: 27px;
```
Лишние отступы

pavel-niukalo commented on styles/styles.css:
```diff
@@ -503,18 +528,62 @@ body {
-	width: 560px;
+	width: 530px;
+	position: absolute;
+    justify-content: space-between;
+    top: 192px;
```
Стоит везде проверить отступы

pavel-niukalo commented on styles/styles.css:
```diff
@@ -803,28 +905,37 @@ body {
 	margin: 0 0 0 auto;
 }
 
+main-footer-phone-number-link {
```
Точка пропущена - браузер думает, что это тег

mis3rably committed:
Исправлены ошибки

Исправлено поведение focus-visible у некоторых элементов, hover у ссылок в header, стилизация кнопки текущего слайдера в слайдере

mis3rably committed:
Исправлен декоративный элемент в header

mis3rably committed:
Добавлены состояния на формы и чекбоксы

pavel-niukalo commented on catalog.html:
```diff
@@ -14,57 +15,90 @@
+              </a>
+            </li>
+            <li class="main-header-navigation-item flex">
+              <a href="#" class="main-header-navigation-link state-default flex">
+                <svg class="main-header-navigation-link-login-shopping-list-icon" width="16" height="14" fill="none"
+                  xmlns="http://www.w3.org/2000/svg">
+                  <path
+                    d="m15.3 5-3.5.2L8.6.3C8.5.1 8.2 0 8 0c-.2 0-.5.1-.6.3L4.2 5.2H.7c-.4 0-.7.3-.7.7v.2l1.9 6.8c.2.7.7 1.1 1.4 1.1h9.4c.7 0 1.2-.4 1.4-1.1L16 6v-.2c0-.4-.3-.8-.7-.8ZM8 1.9l2.2 3.2H5.9L8 1.9Z"
+                    fill="#000">
+                </svg>
+                Корзина
+                <span class="cart-amount">2</span>
+              </a>
```
Фокус в корзине немного вылезает за цифру

Макет:

![image](https://user-images.githubusercontent.com/52506235/199051588-0d3d91ba-9c32-4487-bc89-f4e14ac1d83f.png)

Сейчас:

![image](https://user-images.githubusercontent.com/52506235/199051643-e47eceab-d4b2-4d72-8d26-f7b60037fdaa.png)

mis3rably replied:
Исправил

pavel-niukalo commented on catalog.html:
```diff
@@ -14,57 +15,90 @@
+              </a>
+            </li>
+          </ul>
+          <div class="main-header-page-navigation-menu flex">
+            <a href="catalog.html" class="main-header-page-navigation-button-link state-default flex">Каталог товаров</a>
+            <button class="main-header-page-navigation-button state-default" type="button">
+              <h2 class="visually-hidden">Открыть меню каталога.</h2>
+              <svg width="50" height="51" viewBox="0 0 50 51" fill="none" xmlns="http://www.w3.org/2000/svg">
+                <circle opacity=".3" cx="25" cy="25.8" r="24" stroke="#000" stroke-width="2" />
+                <g opacity=".3" clip-path="url(#a)">
+                  <path d="M33 24.8h-7v-7h-2v7h-7v2h7v7h2v-7h7v-2Z" fill="#000" />
+                </g>
+                <defs>
+                  <clipPath id="a">
+                    <path fill="#fff" transform="translate(17 17.8)" d="M0 0h16v16H0z" />
+                  </clipPath>
+                </defs>
+              </svg>
+            </button>
```
Тут по стайлгайду у нас идет как будто один элемент

![image](https://user-images.githubusercontent.com/52506235/199052557-4a31f8b9-3218-4258-be9a-8914898ba379.png)

А сейчас они по-отдельности фокусируются

![image](https://user-images.githubusercontent.com/52506235/199052903-15038ed4-f03e-4f9c-a621-db82e3143ae2.png)

Предлагаю как реализовать:
у нас будет ссылка, а картинку просто рядом положить, не через кнопку. Таким образом иконка плюсика у нас будет своего рода декором, без интерактива.

А показ подменю как раз завяжем на фокус и ховер

```html
<a>
  Каталог товаров
  <svg></svg>
</a>
```

mis3rably replied:
Исправил

pavel-niukalo commented on catalog.html:
```diff
@@ -14,57 +15,90 @@
+                  xmlns="http://www.w3.org/2000/svg">
+                  <path
+                    d="M0 12h13a6.3 6.3 0 0 0-4.7-5.5c1-.6 1.7-1.7 1.7-3C10 1.6 8.4 0 6.5 0A3.5 3.5 0 0 0 3 3.5c0 1.3.7 2.4 1.7 3A6.3 6.3 0 0 0 0 12Z"
+                    fill="#000">
```
Нужно поправить ошибки валидатора

![image](https://user-images.githubusercontent.com/52506235/199055243-4cbeaec2-ae6f-48a8-89fd-153e0404b18e.png)

mis3rably replied:
Исправил

pavel-niukalo commented on catalog.html:
```diff
@@ -136,134 +170,189 @@ <h2 class="visually-hidden">Список моноподов для селфи с
                 <ul class="pagination-list list flex">
                   <li class="pagination-item flex">
-                    <a class="pagination-link pagination-current">1</a>
+                    <button class="pagination-link pagination-current" disabled>
+                      <span class="pagination-link-content">1</span>
+                    </button>
```
1. `span` выглядит лишним, можно сразу стилизовать кнопку
2. Лучше вернуть ссылку, как было. Потому что если у нас например не будет `js`, то мы сможем хотя бы перекинуть на страницу с другими товарами

mis3rably replied:
Исправил

pavel-niukalo commented on index.html:
```diff
@@ -14,68 +15,101 @@
+                Сравнить
+              </a>
+            </li>
+            <li class="main-header-navigation-item flex">
+              <a href="#" class="main-header-navigation-link state-default flex">
+                <svg class="main-header-navigation-link-login-shopping-list-icon" width="16" height="14" fill="none"
+                  xmlns="http://www.w3.org/2000/svg">
+                  <path
+                    d="m15.3 5-3.5.2L8.6.3C8.5.1 8.2 0 8 0c-.2 0-.5.1-.6.3L4.2 5.2H.7c-.4 0-.7.3-.7.7v.2l1.9 6.8c.2.7.7 1.1 1.4 1.1h9.4c.7 0 1.2-.4 1.4-1.1L16 6v-.2c0-.4-.3-.8-.7-.8ZM8 1.9l2.2 3.2H5.9L8 1.9Z"
+                    fill="#000">
+                </svg>
+                Корзина
+                <span class="cart-amount">2</span>
+              </a>
+            </li>
+          </ul>
+          <div class="main-header-page-navigation-menu flex">
+            <a href="catalog.html" class="main-header-page-navigation-button-link state-default flex">Каталог товаров</a>
+            <button class="main-header-page-navigation-button state-default" type="button">
+              <h2 class="visually-hidden">Открыть меню каталога.</h2>
+              <svg width="50" height="51" viewBox="0 0 50 51" fill="none" xmlns="http://www.w3.org/2000/svg">
+                <circle opacity=".3" cx="25" cy="25.8" r="24" stroke="#000" stroke-width="2" />
+                <g opacity=".3" clip-path="url(#a)">
+                  <path d="M33 24.8h-7v-7h-2v7h-7v2h7v7h2v-7h7v-2Z" fill="#000" />
+                </g>
+                <defs>
+                  <clipPath id="a">
+                    <path fill="#fff" transform="translate(17 17.8)" d="M0 0h16v16H0z" />
+                  </clipPath>
+                </defs>
+              </svg>
+            </button>
```
Здесь так же как выше писал

mis3rably replied:
Это единственное что я не понял что нужно исправить)

pavel-niukalo replied:
Тут я имел ввиду, как в `index.html`, вставить свг-шкой

pavel-niukalo commented on index.html:
```diff
@@ -88,13 +122,15 @@ <h2 class="product-showcase-heading title">Порхает как бабочка,
+          <section class="product-showcase-slider-item flex">
             <h2 class="visually-hidden">Слайдер с демонстрацией товара и его описанием.</h2>
-            <img class="product-showcase-img" src="images/slider-band.png" width="560" height="560" alt="Фотография товара">
+            <img class="product-showcase-img" src="images/slider-band.png" width="560" height="560"
```
Пройдись по всем картинкам и интерактивным элементам. В конце их описаний надо добавить точку

Вот критерии - [1](https://up.htmlacademy.ru/htmlcss/35/criteries#ALLY-01), [2](https://up.htmlacademy.ru/htmlcss/35/criteries#ALLY-02)

mis3rably replied:
Проверил, должно быть всё в порядке

mis3rably committed:
Исправлены ошибки

mis3rably commented on index.html:
```diff
@@ -14,68 +15,161 @@
+            </li>
+          </ul>
+          <div class="main-header-page-navigation-menu flex">
+            <a href="catalog.html" class="main-header-page-navigation-button-link state-default flex">Каталог
+              товаров
+              <svg class="main-header-page-navigation-button-icon" width="50" height="51" viewBox="0 0 50 51" fill="none" xmlns="http://www.w3.org/2000/svg"><circle cx="25" cy="25.8" r="24" stroke="#000" stroke-width="2"/><g clip-path="url(#a)"><path d="M33 24.8h-7v-7h-2v7h-7v2h7v7h2v-7h7v-2Z" fill="#000"/></g><defs><clipPath id="a"><path fill="#fff" d="M17 17.8h16v16H17z"/></clipPath></defs></sv ...
+            </a>
```
Не знаю как сделать в соответствии с Стайлгайдом, получается кнопка при открытии меню с помощью мышки или таба меняется с плюс на минус. Как это сделать если картинка сделана как инлайн свг

pavel-niukalo replied:
Хм, думал, что у нас там плюс нарисован как два отдельных тега (`path`). В таком случае можно было бы навесить класс, и скрывать одну черточку.

Можно попробовать поиграться в самой фигме и нарисовать самому такую конструкцию, либо как вариант еще, экспорнуть не +, а -, и псевдоэлементом нарисовать вертикальную палку, которую, опять же, скрывать

mis3rably replied:
Понял, буду пробовать реализовать с помощью псевдоэлемента

mis3rably committed:
Добавлен новый класс

Добавлен класс new-icon для возможности добавить иконку NEW в карточку товара

pavel-niukalo commented on index.html:
```diff
@@ -405,11 +471,50 @@ <h2 class="subscription-schedule title">Подпишитесь на рассыл
           </a>
         </div>
       </footer>
+      <div class="modal-container modal-close">
+          <section class="modal-content flex">
```
Лишний отступ получился

pavel-niukalo commented on index.html:
```diff
@@ -405,11 +471,50 @@ <h2 class="subscription-schedule title">Подпишитесь на рассыл
+                  Постараемся найти, сколько нужно. Привезём сколько получится по требованиям к провозу багажа.
+                </p>
+                <p class="modal-input-number-button-container flex">
+                  <button class="modal-input-number-button input-decrease"></button>
```
Здесь нужно скрытую подпись добавить

pavel-niukalo commented on index.html:
```diff
@@ -405,11 +471,50 @@ <h2 class="subscription-schedule title">Подпишитесь на рассыл
+                <p class="modal-input-number-button-container flex">
+                  <button class="modal-input-number-button input-decrease"></button>
+                  <input type="number" class="modal-input modal-input-number">
+                  <button class="modal-input-number-button input-increase"></button>
```
Тут так же

pavel-niukalo commented on index.html:
```diff
@@ -405,11 +471,50 @@ <h2 class="subscription-schedule title">Подпишитесь на рассыл
+                  <input type="number" class="modal-input modal-input-number">
+                  <button class="modal-input-number-button input-increase"></button>
+                </p>
+              </div>
+              </fieldset>
```
Тут тоже отступы надо поправить

pavel-niukalo commented on catalog.html:
```diff
@@ -39,40 +40,97 @@ <h1 class="visually-hidden">Интернет-магазин «Девайс».</h
-                </defs>
-              </svg>
-            </button>
+            <a href="catalog.html" class="main-header-page-navigation-button-link state-default flex">Каталог
+              товаров</a>
```
Тут нам тоже надо добавить иконку

pavel-niukalo commented on styles/styles.css:
```diff
@@ -247,7 +310,53 @@ body {
+  line-height: 21px;
+  background-color: transparent;
+  border: 0;
+  border-bottom: 3px solid #c4c4c4;
+  width: 557px;
+  padding: 0;
+  padding-left: 16px;
+  box-sizing: border-box;
+}
+
+.subscription-form-email-input:focus-visible {
+  background-color: #DCDCDC;
```
Для единообразия давай писать цвета в одном регистре - либо верхнем, либо нижнем. То есть сейчас идет смешение - `#c4c4c4` записан в нижнем, а `#DCDCDC` уже в верхнем

mis3rably replied:
Я хотел окончательно это сделать с linthtml в 10 разделе

pavel-niukalo replied:
Тогда ладно, оставим на 10 раздел

pavel-niukalo commented on styles/styles.css:
```diff
@@ -332,13 +463,39 @@ body {
+  appearance: none;
+  width: 146px;
+  padding-left: 1px;
+  background-image: url(../images/catalog-select-arrow.svg);
```
Кавычки в `url` стоит добавить, ждя надежности

mis3rably replied:
Простите, всё время забываю(

pavel-niukalo commented on styles/styles.css:
```diff
@@ -1012,114 +1245,427 @@ body {
+  transform: translate(-50%,-50%);
+}
+
+.control-input:checked + .control-input-checkbox::before {
+  opacity: 1;
+}
+
+.control-input:checked + .control-input-checkbox {
+  border: 2px solid #000000;
+}
```
Так же надо добавить состояние `disabled` для чекбоксов и радио. Если на макете нет, надо самому придумать

mis3rably replied:
А active состояние (подчеркивание у текущей ссылки в хэдере) нужно добавить во время нажатия или в виде отдельного класса который будет добавляться когда пользователь переходит на эту страницу?

pavel-niukalo replied:
Вот это?

![image](https://user-images.githubusercontent.com/52506235/200178629-4cf62c7e-ec7c-4740-a931-60abc826853a.png)

Тут это отдельный класс, который будет показывать активную страницу. `:active` тут не подходит

mis3rably replied:
Да, я имел ввиду это. Буду исправлять

mis3rably replied:
Исправил, также добавил disabled состояния

mis3rably committed:
Исправлены ошибки

Исправлены отступы, добавлен класс для активного состояния у ссылок

mis3rably committed:
Добавлены состояния для disabled

Добавлены состояния disabled для чекбоксов, радио и кнопок

mis3rably committed:
Исправлена вёрстка второго слайдера
